plutus-chain-index/src/Plutus/ChainIndex/Tx.hs

```diff
@@ -1,23 +1,91 @@
+    , citxMintingPolicies
+    , citxValidators
+    , citxTxId
     ) where
 
-import           Data.Map (Map)
-import           Data.Set (Set)
-import           Ledger   (Datum, DatumHash, OnChainTx (..), SlotRange, Tx (..), TxId, TxIn, TxOut, txId)
+import           Control.Lens (makeLenses)
+import           Data.Map     (Map)
+import qualified Data.Map     as Map
+import           Data.Set     (Set)
+import qualified Data.Set     as Set
+import           GHC.Generics (Generic)
+import           Ledger       (Datum, DatumHash, MintingPolicy, MintingPolicyHash, OnChainTx (..), Redeemer (..),
+                               RedeemerHash, SlotRange, Tx (..), TxId, TxIn (txInType), TxInType (..), TxOut,
+                               TxOutRef (..), Validator, ValidatorHash, datumHash, mintingPolicyHash, redeemerHash,
+                               txId, validatorHash)
 
 data ChainIndexTx = ChainIndexTx {
-    citxId         :: TxId,
-    citxInputs     :: Set TxIn,
-    citxOutputs    :: [TxOut],
-    citxValidRange :: !SlotRange,
-    citxData       :: Map DatumHash Datum
-    } deriving (Show, Eq)
+    _citxTxId            :: TxId,
+    _citxInputs          :: Set TxIn,
+    _citxOutputs         :: [TxOut],
+    _citxValidRange      :: !SlotRange,
+    _citxData            :: Map DatumHash Datum,
+    _citxRedeemers       :: Map RedeemerHash Redeemer,
+    _citxMintingPolicies :: Map MintingPolicyHash MintingPolicy,
+    _citxValidators      :: Map ValidatorHash Validator
+    } deriving (Show, Eq, Generic)
+
+makeLenses ''ChainIndexTx
+
+txOutRefs :: ChainIndexTx -> [(TxOut, TxOutRef)]
+txOutRefs ChainIndexTx{_citxTxId, _citxOutputs} =
+    map (\(output, idx) -> (output, TxOutRef _citxTxId idx)) $ zip _citxOutputs [0..]
 
 fromOnChainTx :: OnChainTx -> ChainIndexTx
-fromOnChainTx (Valid tx@Tx{..})   = ChainIndexTx (txId tx) txInputs txOutputs txValidRange txData
-fromOnChainTx (Invalid tx@Tx{..}) = ChainIndexTx (txId tx) txCollateral mempty txValidRange mempty
+fromOnChainTx = \case
+    Valid tx@Tx{txInputs, txOutputs, txValidRange, txData, txMintScripts} ->
+        let (validatorHashes, otherDataHashes, redeemers) = validators txInputs in
+        ChainIndexTx
+            { _citxTxId = txId tx
+            , _citxInputs = txInputs
+            , _citxOutputs = txOutputs
+            , _citxValidRange = txValidRange
+            , _citxData = txData <> otherDataHashes
+            , _citxRedeemers = redeemers
+            , _citxMintingPolicies = mintingPolicies txMintScripts
+            , _citxValidators = validatorHashes
+            }
+    Invalid tx@Tx{txCollateral, txValidRange, txData, txInputs, txMintScripts} ->
+        let (validatorHashes, otherDataHashes, redeemers) = validators txInputs in
+        ChainIndexTx
+            { _citxTxId = txId tx
+            , _citxInputs = txCollateral
+            , _citxOutputs = mempty
+            , _citxValidRange = txValidRange
+            , _citxData = txData <> otherDataHashes
```

Comment:
not sure it if was intentional, but this used to be `mempty`?

Reply:
Good spot, but this is intentional - the "disk state" part of the chain index is basically a content addressable store for pieces of Plutus data so, so we can include the datum hashes here even if the outputs don't make it into the ledger.